### PR TITLE
Update CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by rrzhang139**
### Description

Updating contribution guidelines for more information.

### Issue Link

N/A

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- Best Coding Practices Section instructing streamlined guidelines how to Open PRs, submit issues
- Adding Tests
- FAQs

### Rationale

More thorough documentation on how to contribute, as well as more resources/standards on how to contribute. Previous guidelines are not as streamlined or structured as this one. 

### How Has This Been Tested?

N/A

### CODE REVIEW

**Comment by rrzhang139** [comment]
@ChihChengLiang thoughts on new contributor guide?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,27 +1,92 @@
...
+    ```
+    make test
+    make clippy
+    ```
+    
+7. Once you made the changes, give the commit message a sufficient description of the change. The message content can be brief, including only enough details to summarize the purpose of the commit. In other instances, more in-depth explanations about how the commit accomplishes its goals are better suited as comments in the code.
+    - ******************NOTE:****************** The commits in a pull request should be organized in a manner where each commit represents a small, coherent step towards the pull request's overall objective. The structure of the pull request should facilitate the reviewer in easily tracking and understanding each change made.
+8. You may run into unnecessary commits, such as merge conflicts or minor refactorings. In that case, always `git rebase -i`  and [squash](https://www.git-tower.com/learn/git/faq/git-squash) all your commits to one commit. 
+9. Please do not merge master into your branch as you develop your pull request; instead, rebase your branch on top of the latest master if your pull request branch is long-lived.
+    1. How to regularly update your PR with master:
```
Comment (ChihChengLiang) [inline CONTRIBUTING.md:30]: If we assume the contributor needs this detailed git command guide, then squashing commits might be difficult for them. We can make the commits squashing optional.

**Review by ChihChengLiang — COMMENTED** [review]
Sorry, I thought the PR was at the drafting stage.
I left a comment on squashing.

**Review by ChihChengLiang — APPROVED** [review]
LGTM.

**Comment by ChihChengLiang** [comment]
Would be great if @ed255 or @aguzmant103 can give it a look.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,27 +1,92 @@
...
+
+First take a look at our [specs](https://github.com/privacy-scaling-explorations/zkevm-specs), [auditing videos](https://www.youtube.com/watch?v=HhHTho2QZa4), and [zkevm docs](https://github.com/privacy-scaling-explorations/zkevm-docs). It’s very important that you understand at a high-level how the ZK-EVM architecture and all of its components work.
+
+## Best Coding Practices
+
+### Creating PRs
+
+1. Fork the zkevm or specs repos.
+2. Write a simple markdown spec doc about the circuit PR and send to the `zkevm-specs` repo.
+3. For PRs that change significantly the codebase or the way on which some procedure is done (test-suite, cross-crate structures, module refactors) please [file an issue](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/new/choose) and wait until this feature/proposal gets approved by the repo maintainers.
```
Comment (ed255) [inline CONTRIBUTING.md:15]: I would rephrase this a bit: We encourage everyone to open an issue before getting to work on a PR.  This will be a must for "PRs that change significantly the codebase or the way on which some procedure ..."

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,27 +1,92 @@
...
+4. Send the circuit PR to the `zkevm-circuits` repo. Make sure you give a clear title and succinct description of the PR and the area you are changing.
```
Comment (ed255) [inline CONTRIBUTING.md:16]: Following the steps chronologically, I think this point should go after current point 7.

File: CONTRIBUTING.md (hunk 1 above)
Comment (ed255) [reply CONTRIBUTING.md:30]: We must also consider that currently for outdated branches in a PR, github shows a button called "Update branch" that merges main into the branch (instead of doing the rebase).  This seems to be very convenient for reviewers who want to merge a PR from a contributor without write access; because after an approval they can just click, wait for tests, and merge.  Maybe we could rephrase points 8 and 9 as suggestions?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,27 +1,92 @@
...
+9. Please do not merge master into your branch as you develop your pull request; instead, rebase your branch on top of the latest master if your pull request branch is long-lived.
+    1. How to regularly update your PR with master:
+        
+        ```
+        git checkout main
+        git pull
+        git checkout my/branch
+        git merge main
+        git rebase -i // squash commits!
+        git push
```
Comment (ed255) [inline CONTRIBUTING.md:38]:
```suggestion
        git rebase main
        git push -f
```
I think these steps are simpler and don't require an interactive rebase

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,27 +1,92 @@
...
+    ```
+    make test
+    make clippy
+    ```
+    
+7. Once you made the changes, give the commit message a sufficient description of the change. The message content can be brief, including only enough details to summarize the purpose of the commit. In other instances, more in-depth explanations about how the commit accomplishes its goals are better suited as comments in the code.
+    - ******************NOTE:****************** The commits in a pull request should be organized in a manner where each commit represents a small, coherent step towards the pull request's overall objective. The structure of the pull request should facilitate the reviewer in easily tracking and understanding each change made.
+8. [OPTIONAL] You may run into unnecessary commits, such as merge conflicts or minor refactorings. In that case, always `git rebase -i`  and [squash](https://www.git-tower.com/learn/git/faq/git-squash) all your commits to one commit. 
+9. Please do not merge master into your branch as you develop your pull request; instead, rebase your branch on top of the latest master if your pull request branch is long-lived.
+    1. How to regularly update your PR with master:
```
Comment (ed255) [inline CONTRIBUTING.md:30]:
```suggestion
9. Please do not merge main into your branch as you develop your pull request; instead, rebase your branch on top of the latest main if your pull request branch is long-lived.
    1. How to regularly update your PR with main:
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,27 +1,92 @@
...
+        git checkout my/branch
+        git merge main
+        git rebase -i // squash commits!
+        git push
+        ```
+        
+
+### Submitting Bug Issues
+
+1. If you identified a bug, please create a new issue or discuss it in our forums.
```
Comment (ed255) [inline CONTRIBUTING.md:44]: Which forums?  Is this a reference to https://github.com/privacy-scaling-explorations/zkevm-circuits/discussions ?  If so, could you add a link?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,27 +1,92 @@
...
+        git rebase -i // squash commits!
+        git push
+        ```
+        
+
+### Submitting Bug Issues
+
+1. If you identified a bug, please create a new issue or discuss it in our forums. 
+2. In the description, provide the following:
+    - What command(s) is the bug in?
```
Comment (ed255) [inline CONTRIBUTING.md:46]: Does this apply in our case? zkevm-circuits is more a library than a program, so usually there are no commands.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,27 +1,92 @@
...
+- **Integration tests**: For general purpose, far reaching functionality, integration tests should be added. The best way to add a new integration test is to look at existing ones and follow the style.
 
 ## General considerations
-- Make atomic PRs. Focused on a single feature or topic. Don't mix different topics or issues in the same PR if it can be avoided.
-- For PRs that change significantly the codebase or the way on which some procedure is done (test-suite, cross-crate structures, module refactors) please file an issue and wait until this feature/proposal gets approved by the repo maintainers.
-- Don't set PRs as ready for review until they're not completed from the dev perspective and also are up-to-date with `main`/`master`. Also, if any changes are requested, bring back the PR to `Draft` until it's ready again with the aforementioned requirements.
+
+- Don't set PRs as ready for review until they're not completed from the dev perspective and also are up-to-date with `main`/`master`. Also, if any changes are requested, bring back the PR to `Draft` until it's ready again with the aforementioned requirements.
```
Comment (ed255) [inline CONTRIBUTING.md:70]: > Also, if any changes are requested, bring back the PR to `Draft` until it's ready again with the aforementioned requirements.

We're not usually doing this.  I think having unresolved requests from a reviewer, or having the PR marked as changes requested is enough to signal that the PR is not ready to be merged and the PR submitter has work to do.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,27 +1,92 @@
...
+## FAQs
+
+Q: Do I need permissions on PSE org to create issues or PRs?
+
+A: You don’t need permissions from PSE, can just create an issue or propose a PR and we will assign you issues or review/approve/merge PRs. 
+
+Q: Is there a forum for discussing different issues or features where I can get started?
 
-- Try to include `bus-mapping` + `zkevm-circuits` opcode implementations in the same PR, but never mixing more than one Opcode at a time.
+A: There’s the issues section in zkevm repos, but for more nuanced discussion join our [Github Discussions](https://github.com/privacy-scaling-explorations/zkevm-circuits/discussions) or [discord](https://discord.gg/BvcMkbKN).
```
Comment (ed255) [inline CONTRIBUTING.md:92]: I tried following the discord link and it tells me "Invite Invalid"

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for addressing the requests :)
